Separate concretization of build dependencies (coupled)

alalazo opened the pull request:
closes #35739 
closes #34889 
closes #27835
closes #38587

:warning: **Attempt to get separate concretization of build dependencies by using a more complex ASP encoding** :warning: 

### Tasks
- [x] Mark all the rules that are associated with packages (1e8ff49f382bb4181aeef257f513b5ffc31b39ec)
- [x] Introduce `node(ID, Package)` nested facts, but still allow only for `ID==0` (a87e2a746e2520614011b85773457ca67e26e659)
- [x] Add choice rules to allow clingo to generate nodes, and introduce "unification sets"
- [x] Allow different concretization modes for separate concretization
- [x] Allow multiple nodes per package
- [x] Make sure that `conflicts` refer to direct dependency or self only
- [x] Add a `vendors` directive for cases where a package `conflicts` with a random other because it exposes its own internal version of it

alalazo committed:
Transform many package related facts to use a nested function

Each fact that is deduced from package rules, and start with
a bare package atom, is transformed into a "facts" atom containing
a nested function.

For instance we transformed

  version_declared(Package, ...) -> facts(Package, version_declared(...))

This allows us to clearly mark facts that represent a rule on the package,
and will be of help later when we'll have to distinguish the cases where
the atom "Package" is being used referred to package rules and not to a
node in the DAG.

alalazo committed:
Remove useless rule

The version_equivalent fact was deleted in #36347,
but the corresponding rule was not removed.

alalazo committed:
Remove unneeded #defined directives

alalazo committed:
Rework the encoding to introduce node(ID, Package) nested facts

So far the encoding has a single ID per package, i.e. all the
facts will be node(0, Package). This will prepare the stage for
extending this logic and having multiple nodes from the same
package in a DAG.

alalazo committed:
Allow clingo to generate edges

alalazo committed:
Introduce unification sets

Unification sets are possibly overlapping sets of nodes that
might contain at most a single configuration for any package.

alalazo commented:
Comparing the transformation of the encoding to prepare it for the introduction of separate concretization of build dependencies:

[radiuss.8669f3e3f153de172329bc3394991b64f530665c.csv](https://github.com/spack/spack/files/11788438/radiuss.8669f3e3f153de172329bc3394991b64f530665c.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/11788439/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11788440/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/e98d5589-8be6-4594-8a64-35b646d3c134)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       17.5607 |                  19.7323 |     -12.3665  |
| ascent        |                       36.1886 |                  39.304  |      -8.60879 |
| axom          |                       34.9225 |                  39.4993 |     -13.1058  |
| blt           |                       16.1301 |                  18.0389 |     -11.8335  |
| caliper       |                       17.4336 |                  20.0213 |     -14.8432  |
| care          |                       18.0698 |                  21.0498 |     -16.4921  |
| chai          |                       17.3742 |                  20.0581 |     -15.4472  |
| conduit       |                       17.8551 |                  20.7511 |     -16.2189  |
| dihydrogen    |                       19.4309 |                  21.2052 |      -9.13093 |
| flux-core     |                       17.9489 |                  20.7459 |     -15.5829  |
| flux-sched    |                       20.5557 |                  20.9913 |      -2.11928 |
| flux-security |                       15.8155 |                  19.1619 |     -21.1594  |
| glvis         |                       32.9164 |                  40.886  |     -24.2118  |
| hydrogen      |                       17.4141 |                  20.4855 |     -17.6374  |
| hypre         |                       18.8567 |                  22.0955 |     -17.1758  |
| lbann         |                       34.8106 |                  43.0737 |     -23.7372  |
| lvarray       |                       22.4491 |                  25.1766 |     -12.1496  |
| mfem          |                       31.527  |                  36.5541 |     -15.9453  |
| py-hatchet    |                       20.7295 |                  27.4639 |     -32.487   |
| py-maestrowf  |                       17.9352 |                  20.6887 |     -15.3526  |
| py-merlin     |                       18.8993 |                  23.417  |     -23.9041  |
| py-shroud     |                       15.7452 |                  18.3862 |     -16.7737  |
| raja          |                       16.5619 |                  19.5622 |     -18.1154  |
| samrai        |                       16.1204 |                  19.3522 |     -20.0475  |
| scr           |                       18.0463 |                  21.1855 |     -17.395   |
| sundials      |                       27.277  |                  32.502  |     -19.1554  |
| umpire        |                       16.1656 |                  18.6915 |     -15.6253  |
| visit         |                       34.2725 |                  44.3729 |     -29.4709  |
| xbraid        |                       16.4195 |                  19.0473 |     -16.0039  |
| zfp           |                       15.7581 |                  18.427  |     -16.9365  |
</details>

So far only a single node is allowed, and the runtime is expected further slow-downs as we introduce separate concretization of build dependencies.[^1]

[^1]: Note, this new encoding hasn't been heavily optimized yet, so we might still speed it up by a few %

alalazo committed:
Parametrize all the logic program for multiple nodes

Introduce the concept of "condition sets", i.e. the set of packages on which
a package can require / impose conditions. This currently maps to the link/run
sub-dag of each package + its direct build dependencies.

Parametrize the "condition" and "requirement" logic to multiple nodes.

alalazo committed:
Reduce the dependency types in a solve

Count the maximum number of nodes based on dependency types

alalazo committed:
Tweak a unit test by allowing a different type of exception to be raised

alalazo commented:
This is a plot of the slow-down due to the more complex encoding in https://github.com/spack/spack/pull/38447/commits/9742a8eb4271da4af77c47a92494d9da5513d499 and https://github.com/spack/spack/pull/38447/commits/4285810818d44848690afeaddc63325777428444 to map nodes to packages. 

Both plots have been obtained allowing a single nodes per package, and they reflect the _effect of the added complexity in the encoding_.

develop: https://github.com/spack/spack/pull/38447/commits/4285810818d44848690afeaddc63325777428444
PR: d761d926a82e8ed7a23c54d8ce608760da764f71

[radiuss.d761d926a82e8ed7a23c54d8ce608760da764f71.csv](https://github.com/spack/spack/files/11824486/radiuss.d761d926a82e8ed7a23c54d8ce608760da764f71.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/11824487/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11824488/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/29c0f970-74c5-472f-9cfc-337f5952f813)


<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.7819 |                  40.5285 |     -141.502  |
| ascent        |                       35.1049 |                  86.5947 |     -146.674  |
| axom          |                       34.0034 |                  82.5838 |     -142.869  |
| blt           |                       15.6801 |                  26.2905 |      -67.6682 |
| caliper       |                       16.8628 |                  40.4477 |     -139.863  |
| care          |                       18.0986 |                  29.5529 |      -63.2886 |
| chai          |                       17.1099 |                  32.3305 |      -88.958  |
| conduit       |                       17.8418 |                  33.1957 |      -86.056  |
| dihydrogen    |                       18.1802 |                  41.8569 |     -130.233  |
| flux-core     |                       17.4456 |                  35.2627 |     -102.13   |
| flux-sched    |                       17.1557 |                  44.0192 |     -156.586  |
| flux-security |                       15.7154 |                  19.9079 |      -26.6774 |
| glvis         |                       33.0898 |                  60.752  |      -83.5972 |
| hydrogen      |                       17.2905 |                  44.5871 |     -157.87   |
| hypre         |                       18.5722 |                  59.2684 |     -219.125  |
| lbann         |                       35.1247 |                  82.9198 |     -136.073  |
| lvarray       |                       22.7947 |                  35.4677 |      -55.5963 |
| mfem          |                       31.6893 |                  65.859  |     -107.827  |
| py-hatchet    |                       21.0032 |                  52.8454 |     -151.607  |
| py-maestrowf  |                       17.6902 |                  27.2252 |      -53.9002 |
| py-merlin     |                       19.3176 |                  50.8648 |     -163.308  |
| py-shroud     |                       16.1795 |                  24.5725 |      -51.8748 |
| raja          |                       16.3333 |                  29.1512 |      -78.4773 |
| samrai        |                       16.0283 |                  36.7549 |     -129.313  |
| scr           |                       18.1403 |                  38.4019 |     -111.694  |
| sundials      |                       27.5532 |                  68.381  |     -148.178  |
| umpire        |                       16.6091 |                  27.3682 |      -64.7778 |
| visit         |                       34.2922 |                  68.782  |     -100.576  |
| xbraid        |                       16.1758 |                  49.976  |     -208.956  |
| zfp           |                       15.8366 |                  25.6672 |      -62.0758 |

</details>

The encoding passed a first sanity check, i.e. it passes the unit tests under the same conditions of the previous encoding. Now we need to recover performance. Three possible step forward:
1. Tune the heuristic for this new encoding
2. Deduplicate conditions and impositions from the directives
3. Reduce the number of "possible dependencies" in a build

alalazo committed:
ASP-based solver: do not optimize on known dimensions

All the solution modes we use imply that we have to solve for all
the literals, except for "when possible".

Here we remove a minimization on the number of literals not
solved, and emit directly a fact when a literal *has* to be
solved.

alalazo committed:
Extract a function to emit variant rules

alalazo committed:
Deduplicate trigger and effect conditions in packages

This refactor introduces extra indices for triggers and
effect of a condition, so that the corresponding clauses
are evaluated once for every condition they apply to.

alalazo committed:
Make cycle detection optional, to speed-up grounding and solving

alalazo committed:
Reduce the number of unification sets to only two

alalazo commented:
Compared to https://github.com/spack/spack/pull/38447#issuecomment-1601614061, this is the effect of the following optimizations:
1. 10129b4 (remove the optimization on `literal`s if we know we have to solve for all of them)
2. 8499bc1 (deduplicate condition triggers and effects)
3. fa060e5 (use cycle detection in ASP only as a fall-back)
4. 8b845bd (reduce the number of unification sets)

As before, I am testing against develop under the same conditions, so I allow just one node per package.

develop: https://github.com/spack/spack/commit/4285810818d44848690afeaddc63325777428444
PR: 8b845bd4b157a0f03eb46313b98ef15977960417

[radiuss.8b845bd4b157a0f03eb46313b98ef15977960417.csv](https://github.com/spack/spack/files/11847850/radiuss.8b845bd4b157a0f03eb46313b98ef15977960417.csv)
[radiuss.txt](https://github.com/spack/spack/files/11847851/radiuss.txt)
[radiuss.develop.csv](https://github.com/spack/spack/files/11847852/radiuss.develop.csv)

![radiuss](https://github.com/spack/spack/assets/4199709/576b3b40-6b1a-4987-bc21-82e6b2e767e3)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.7819 |                  16.585  |      1.17332  |
| ascent        |                       35.1049 |                  47.5708 |    -35.5103   |
| axom          |                       34.0034 |                  40.6636 |    -19.5869   |
| blt           |                       15.6801 |                  14.3235 |      8.65187  |
| caliper       |                       16.8628 |                  16.9906 |     -0.757584 |
| care          |                       18.0986 |                  17.7029 |      2.18631  |
| chai          |                       17.1099 |                  16.7097 |      2.33893  |
| conduit       |                       17.8418 |                  17.506  |      1.88211  |
| dihydrogen    |                       18.1802 |                  19.1285 |     -5.21596  |
| flux-core     |                       17.4456 |                  15.1445 |     13.1902   |
| flux-sched    |                       17.1557 |                  15.8922 |      7.36511  |
| flux-security |                       15.7154 |                  14.0325 |     10.7084   |
| glvis         |                       33.0898 |                  39.1442 |    -18.2968   |
| hydrogen      |                       17.2905 |                  17.8254 |     -3.09342  |
| hypre         |                       18.5722 |                  18.8423 |     -1.45477  |
| lbann         |                       35.1247 |                  40.7819 |    -16.1061   |
| lvarray       |                       22.7947 |                  21.4851 |      5.74509  |
| mfem          |                       31.6893 |                  40.8205 |    -28.8147   |
| py-hatchet    |                       21.0032 |                  19.0041 |      9.5177   |
| py-maestrowf  |                       17.6902 |                  16.3678 |      7.47527  |
| py-merlin     |                       19.3176 |                  17.3405 |     10.2349   |
| py-shroud     |                       16.1795 |                  14.9685 |      7.48437  |
| raja          |                       16.3333 |                  15.5686 |      4.68189  |
| samrai        |                       16.0283 |                  15.7052 |      2.016    |
| scr           |                       18.1403 |                  17.2927 |      4.67249  |
| sundials      |                       27.5532 |                  37.3038 |    -35.3884   |
| umpire        |                       16.6091 |                  15.5887 |      6.14373  |
| visit         |                       34.2922 |                  62.791  |    -83.1058   |
| xbraid        |                       16.1758 |                  14.747  |      8.83266  |
| zfp           |                       15.8366 |                  15.0807 |      4.77325  |

</details>

### Notes

- The optimizations done here also apply to `develop`, at least conceptually.
- `visit` is still slower, because it falls back to using cycle detection - since the optimal solution without the integrity constraint *has* cycles[^1]
- Removing cycle detection is a huge win in solving time (now), and an enabler for grounding with multiple nodes

[^1]: In my opinion this is fine, since the issue can be worked around in either spec literals or in package defaults.

alalazo committed:
Construct unification sets on demand, improve heuristic

alalazo committed:
Fix a few bugs in the encoding when imposing constraints on build deps only

These bugs would show up when we try to split nodes by
imposing different targets or different compilers to all
build dependencies.

alalazo commented:
Latest commit is a step towards getting full separation:
- [6779636](https://github.com/spack/spack/pull/38447/commits/67796368c021c5914ce8e832c9019c8812b9fef3) allows each software tagged "`build-tools` " and used as a build dependency to have its own unification set. This keeps the number of unification sets low, but serves the purpose of separating over e.g. cmake, gamke, py-setuptools etc.
- [d340ba6](https://github.com/spack/spack/pull/38447/commits/d340ba67cb3650a78b2f14fcab4f5c6bb89be244) encodes a couple of rules differently, to enable enforcing different targets and / or compilers for specs in the "root" unification set vs. other unification sets

develop: 78f33bc002816dd690047a7a38077b6ac6617d44
PR: d340ba67cb3650a78b2f14fcab4f5c6bb89be244

[radiuss.d340ba67cb3650a78b2f14fcab4f5c6bb89be244.csv](https://github.com/spack/spack/files/11895064/radiuss.d340ba67cb3650a78b2f14fcab4f5c6bb89be244.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/11895065/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11895066/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/ef938957-45f9-4064-ab25-9af880958315)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.3646 |                  16.3962 |     -0.193358 |
| ascent        |                       34.6264 |                  45.9924 |    -32.8246   |
| axom          |                       33.1618 |                  41.8434 |    -26.1795   |
| blt           |                       15.2374 |                  15.0361 |      1.32114  |
| caliper       |                       16.4263 |                  16.8609 |     -2.64528  |
| care          |                       17.7088 |                  18.1668 |     -2.58636  |
| chai          |                       16.774  |                  17.2125 |     -2.61438  |
| conduit       |                       17.3509 |                  18.4839 |     -6.52994  |
| dihydrogen    |                       17.6146 |                  18.8824 |     -7.19764  |
| flux-core     |                       16.7112 |                  15.9129 |      4.77745  |
| flux-sched    |                       17.0186 |                  16.3693 |      3.81521  |
| flux-security |                       15.6706 |                  14.1482 |      9.71494  |
| glvis         |                       32.2706 |                  39.172  |    -21.3861   |
| hydrogen      |                       16.8953 |                  17.442  |     -3.23545  |
| hypre         |                       18.4006 |                  20.379  |    -10.7517   |
| lbann         |                       33.9559 |                  41.896  |    -23.3838   |
| lvarray       |                       22.0815 |                  22.0366 |      0.203414 |
| mfem          |                       30.562  |                  40.4265 |    -32.2768   |
| py-hatchet    |                       20.717  |                  18.8091 |      9.2092   |
| py-maestrowf  |                       17.1667 |                  15.5794 |      9.24636  |
| py-merlin     |                       18.6637 |                  18.1155 |      2.93725  |
| py-shroud     |                       15.6657 |                  14.472  |      7.61973  |
| raja          |                       16.1789 |                  16.0032 |      1.08627  |
| samrai        |                       15.7881 |                  16.0464 |     -1.6363   |
| scr           |                       18.062  |                  17.0265 |      5.73296  |
| sundials      |                       27.449  |                  37.6311 |    -37.0946   |
| umpire        |                       15.8718 |                  15.8215 |      0.317381 |
| visit         |                       33.6443 |                  56.999  |    -69.4164   |
| xbraid        |                       16.1397 |                  16.0133 |      0.783393 |
| zfp           |                       15.5562 |                  14.5621 |      6.39015  |

</details>

alalazo committed:
Add a new configuration option to select among different concretization modes

The "concretizer" section has been extended with a "duplicates:strategy"
attribute, that can take three values:

- "none": only 1 node per package
- "minimal": allow multiple nodes opf specific packages
- "full": allow full duplication for a build tool

alalazo committed:
Add unit-tests for use cases requiring separate concretization of build deps

alalazo committed:
Fix computation of max nodes

alalazo commented:
Adding these rules to a solve (by inserting them in `display.lp`):
```
attr("node_target_set", PackageNode, "x86_64_v4") :- unification_set("root", PackageNode), attr("node", PackageNode).
attr("node_target_set", PackageNode, "x86_64_v2") :- unification_set(SetID, PackageNode), attr("node", PackageNode), SetID != "root".
attr("node_compiler_version_set", PackageNode, "oneapi", "2023.1.0") :- unification_set("root", PackageNode), attr("node", PackageNode).
attr("node_compiler_version_set", PackageNode, "gcc", "9.4.0") :- unification_set(SetID, PackageNode), attr("node", PackageNode), SetID != "root".
:- provider(node(_, Package), node(_, "mpi")), Package != "openmpi".
```
I was able to get a "full separation" of build dependencies wrt link-run dependencies, and obtained the following DAG for trilinos:
![trilinos](https://github.com/spack/spack/assets/4199709/79e8284e-f109-4692-a340-6a3ae529d8a3)

alalazo committed:
Optimize grounding of "can_inherit_flags"

alalazo commented:
Latest results:

develop: https://github.com/spack/spack/commit/78f33bc002816dd690047a7a38077b6ac6617d44
PR: 69371aee145027cfbf07c8b4a3e0fa6e340c4fa5

[radiuss.develop.csv](https://github.com/spack/spack/files/11916955/radiuss.develop.csv)
[radiuss.69371aee145027cfbf07c8b4a3e0fa6e340c4fa5.csv](https://github.com/spack/spack/files/11916956/radiuss.69371aee145027cfbf07c8b4a3e0fa6e340c4fa5.csv)
[radiuss.txt](https://github.com/spack/spack/files/11916957/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/37df8a7d-be35-40ce-9f11-1c529554a512)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.3646 |                  13.5913 |    16.9472    |
| ascent        |                       34.6264 |                  35.1772 |    -1.59076   |
| axom          |                       33.1618 |                  33.1865 |    -0.0745219 |
| blt           |                       15.2374 |                  12.3033 |    19.256     |
| caliper       |                       16.4263 |                  13.641  |    16.9563    |
| care          |                       17.7088 |                  15.4588 |    12.7055    |
| chai          |                       16.774  |                  13.8608 |    17.3673    |
| conduit       |                       17.3509 |                  14.5047 |    16.4039    |
| dihydrogen    |                       17.6146 |                  15.0503 |    14.5578    |
| flux-core     |                       16.7112 |                  12.7888 |    23.4721    |
| flux-sched    |                       17.0186 |                  13.1836 |    22.5345    |
| flux-security |                       15.6706 |                  12.1072 |    22.7397    |
| glvis         |                       32.2706 |                  30.5924 |     5.20021   |
| hydrogen      |                       16.8953 |                  14.2023 |    15.9395    |
| hypre         |                       18.4006 |                  15.834  |    13.9485    |
| lbann         |                       33.9559 |                  32.9092 |     3.0824    |
| lvarray       |                       22.0815 |                  18.0809 |    18.1173    |
| mfem          |                       30.562  |                  33.1487 |    -8.46383   |
| py-hatchet    |                       20.717  |                  15.2254 |    26.5076    |
| py-maestrowf  |                       17.1667 |                  13.5857 |    20.8602    |
| py-merlin     |                       18.6637 |                  14.2717 |    23.5321    |
| py-shroud     |                       15.6657 |                  12.4527 |    20.5102    |
| raja          |                       16.1789 |                  13.1257 |    18.8713    |
| samrai        |                       15.7881 |                  12.6943 |    19.5954    |
| scr           |                       18.062  |                  13.7979 |    23.6078    |
| sundials      |                       27.449  |                  27.1293 |     1.16484   |
| umpire        |                       15.8718 |                  13.1134 |    17.3797    |
| visit         |                       33.6443 |                  80.375  |  -138.896     |
| xbraid        |                       16.1397 |                  13.1302 |    18.6468    |
| zfp           |                       15.5562 |                  12.0939 |    22.257     |
</details>

@tgamblin @becker33 With the exception of `visit`, which takes twice as much as before due to cycle detection, many other specs are ~20% faster in this branch than on `develop` - when we allow a single node per package. 

I am able to solve https://github.com/spack/spack/pull/38447#issuecomment-1614372997 in a "fully separate way" in ~15 seconds by adding the 5 rules you see in the comment (to help looking for a solution). 

I think this PR is ready for a review, to merge `concretize.lp` in `develop` and have a common base for further changes in the model. There is still space for further optimizations, but I would leave them for future PRs.

alalazo committed:
Fix ecp-data-vis-sdk: a package cannot impose constraints on transitive deps

alalazo commented:
@spackbot run pipeline

alalazo committed:
Improve handling of cases with cycles

To avoid paying the cost of setup and of a full grounding again,
move cycle detection into a separate program and check first if
the solution has cycles.

If it has, ground only the integrity constraint preventing cycles
and solve again.

alalazo committed:
We cannot require "mpich" as an mpi provider and ask for openmpi

alalazo commented:
Some change in the packages must have "fixed" the cycle on `visit`, so here are the latest results.

develop: a0b2ca2dde9c891e78908e7ac4915f1b2a553258
PR: https://github.com/spack/spack/pull/38447/commits/1b232fa38abe09dced68b4d70f146aa20cfd3e61

[radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.csv](https://github.com/spack/spack/files/11951319/radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/11951320/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/11951321/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/024a4c60-e60f-4deb-9983-39c38f5abf2a)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [develop, secs.] |   Total time [PR, secs.] |   Speedup [%] |
|---------------|-------------------------------|--------------------------|---------------|
| aluminum      |                       16.9257 |                  13.6296 |      19.4734  |
| ascent        |                       37.0168 |                  35.3102 |       4.61053 |
| axom          |                       33.8625 |                  33.3101 |       1.63131 |
| blt           |                       15.5815 |                  12.103  |      22.3247  |
| caliper       |                       16.996  |                  13.9773 |      17.7612  |
| care          |                       18.3477 |                  15.6307 |      14.8086  |
| chai          |                       17.3217 |                  14.0287 |      19.0108  |
| conduit       |                       17.4397 |                  14.5766 |      16.417   |
| dihydrogen    |                       18.0457 |                  15.1472 |      16.0624  |
| flux-core     |                       17.1302 |                  12.8922 |      24.74    |
| flux-sched    |                       17.6314 |                  12.9195 |      26.7247  |
| flux-security |                       15.7293 |                  12.1357 |      22.8468  |
| glvis         |                       32.7757 |                  31.2987 |       4.50628 |
| hydrogen      |                       17.5115 |                  14.4576 |      17.4393  |
| hypre         |                       19.0606 |                  16.3876 |      14.0235  |
| lbann         |                       34.6988 |                  33.5533 |       3.30106 |
| lvarray       |                       22.4701 |                  18.595  |      17.2455  |
| mfem          |                       31.3069 |                  33.4942 |      -6.98688 |
| py-hatchet    |                       21.3346 |                  15.2888 |      28.3379  |
| py-maestrowf  |                       17.751  |                  13.6104 |      23.3259  |
| py-merlin     |                       19.0097 |                  14.3082 |      24.732   |
| py-shroud     |                       15.9185 |                  12.1884 |      23.433   |
| raja          |                       16.6874 |                  13.4987 |      19.1082  |
| samrai        |                       16.3609 |                  12.7379 |      22.144   |
| scr           |                       18.3752 |                  13.8282 |      24.7453  |
| sundials      |                       27.9783 |                  26.9376 |       3.71961 |
| umpire        |                       16.3472 |                  12.8785 |      21.2194  |
| visit         |                       34.7585 |                  32.3719 |       6.86626 |
| xbraid        |                       16.3976 |                  12.7296 |      22.3692  |
| zfp           |                       15.9595 |                  12.4483 |      22.0007  |

</details>

We are roughly 20% faster than current `develop` on this benchmark, with still some room to improve further in following PRs.

alalazo commented:
This instead is the performance comparison for using the "minimal duplicates" mode that can solve #38671 

PR: https://github.com/spack/spack/commit/1b232fa38abe09dced68b4d70f146aa20cfd3e61

[radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.minimal.csv](https://github.com/spack/spack/files/11951534/radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.minimal.csv)
[radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.csv](https://github.com/spack/spack/files/11951535/radiuss.1b232fa38abe09dced68b4d70f146aa20cfd3e61.csv)
[radiuss.txt](https://github.com/spack/spack/files/11951536/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/53414a81-8b14-43fa-a1ef-4bf38fcb7ba6)

<details>

<summary>Table details</summary>

| Spec solved   |   Total time [PR - no duplicates, secs.] |   Total time [PR - minimal duplicates, secs.] |   Speedup [%] |
|---------------|------------------------------------------|-----------------------------------------------|---------------|
| aluminum      |                                  13.6296 |                                       15.1293 |     -11.003   |
| ascent        |                                  35.3102 |                                       37.7954 |      -7.03831 |
| axom          |                                  33.3101 |                                       35.7682 |      -7.37931 |
| blt           |                                  12.103  |                                       13.484  |     -11.4104  |
| caliper       |                                  13.9773 |                                       15.0289 |      -7.52349 |
| care          |                                  15.6307 |                                       17.0219 |      -8.90091 |
| chai          |                                  14.0287 |                                       15.3392 |      -9.34149 |
| conduit       |                                  14.5766 |                                       16.3482 |     -12.1534  |
| dihydrogen    |                                  15.1472 |                                       16.7969 |     -10.8911  |
| flux-core     |                                  12.8922 |                                       14.1028 |      -9.39014 |
| flux-sched    |                                  12.9195 |                                       14.5485 |     -12.6094  |
| flux-security |                                  12.1357 |                                       13.2021 |      -8.78777 |
| glvis         |                                  31.2987 |                                       33.4607 |      -6.90777 |
| hydrogen      |                                  14.4576 |                                       16.168  |     -11.8308  |
| hypre         |                                  16.3876 |                                       18.2846 |     -11.5757  |
| lbann         |                                  33.5533 |                                       35.5295 |      -5.88945 |
| lvarray       |                                  18.595  |                                       19.5223 |      -4.98675 |
| mfem          |                                  33.4942 |                                       36.431  |      -8.76795 |
| py-hatchet    |                                  15.2888 |                                       16.5183 |      -8.04138 |
| py-maestrowf  |                                  13.6104 |                                       16.8965 |     -24.1443  |
| py-merlin     |                                  14.3082 |                                       15.4326 |      -7.85819 |
| py-shroud     |                                  12.1884 |                                       14.9143 |     -22.3651  |
| raja          |                                  13.4987 |                                       14.5309 |      -7.64614 |
| samrai        |                                  12.7379 |                                       15.1644 |     -19.0493  |
| scr           |                                  13.8282 |                                       15.2941 |     -10.6007  |
| sundials      |                                  26.9376 |                                       29.1909 |      -8.36488 |
| umpire        |                                  12.8785 |                                       14.4575 |     -12.261   |
| visit         |                                  32.3719 |                                       34.3592 |      -6.13918 |
| xbraid        |                                  12.7296 |                                       14.122  |     -10.9384  |
| zfp           |                                  12.4483 |                                       13.9592 |     -12.1373  |

</details>

Slowdown is on average in the 5%-10% range.

alalazo committed:
Removed leftover TODOs

alalazo committed:
ecp-data-viz-sdk: fix building with new encoding

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -549,130 +691,135 @@ error(10, Message) :-
-  attr("variant_propagate", Package, Variant, Value1, Source1),
-  attr("variant_propagate", Package, Variant, Value2, Source2),
-  variant(Package, Variant),
+node_has_variant(node(NodeID, Package), variant(Variant)) :-
```
Why not simplify this to `node_has_variant(node(NodeID, Package), Variant)`? I can't see that the `variant(Variant)` structure is used anywhere.

alalazo replied:
Done in 6be000e

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -918,15 +967,18 @@ def key_fn(version):
             self.gen.fact(
-                fn.version_declared(
-                    pkg.name, declared_version.version, weight, str(declared_version.origin)
+                fn.facts(
```
I think this should be called `pkg_fact()` instead of `facts()`, as the former tells you more about what it means -- it's specifically information that comes from a package file (and thus information that has to be duplicated if we're going to allow two of some package).

`facts()` bothers me because *everything* in the program is really a fact (and it's plural when it's really one fact we're expressing).

I know this is somewhat hypocritical of me because `attr()` is probably worse and should really be renamed to `spec_attr()` or `model_attr()` or something like that.  But seems like that isn't immediately pertinent to this PR -- and I think we should try to have better names.

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/c721ebffb8f7be241e1898e6046b5f33618c34e7

> I know this is somewhat hypocritical of me because [ .. ]

I don't think it is. Code benefits from improved naming, so having pushed a name that is too generic in the past isn't a good reason to let another one get in right now We need to change `attr` in a following PR :slightly_smiling_face:

tgamblin commented on lib/spack/spack/test/concretize.py:
```diff
@@ -152,7 +152,7 @@ class Root(Package):
     depends_on("changing")
 
-    conflicts("changing~foo")
+    conflicts("^changing~foo")
```
Is it not possible to declare a conflict on an arbitrary package anymore?  i.e., do we *have* to add `^`? I see you changed a bunch of packages in `builtin` to do this.

This will break backward compatibility for any packages that did this (remember 2/3 of users have a private package repo, at least according to the survey).

I can see a few ways to resolve this:

1. If we don't care about the `conflicts("package")` case (which I would really take to mean "can't be in a graph with `package` at all") then we could just automatically convert any `conflicts("package")` to `conflicts("^package")` internally, but still accept `conflicts("package")`
2. If we *do* care about `conflicts("package")` meaning something eventually (I think we may someday want to), then we ought to do (1) for now but also warn that it may mean something different someday.  Right now I'm pretty sure that `conflicts("package")` and `conflicts("^package")` mean the same thing.
3. If we just don't care about `conflicts("package")` and want to not allow it (personally this is my least favorite and I'd argue against it) then we could do (2) but warn that it'll be deleted.

What we can't do is make `conflicts("package")` breaking all of a sudden when it wasn't before.

Thoughts?

alalazo replied:
> Is it not possible to declare a conflict on an arbitrary package anymore?

Declaring a conflict on an arbitrary package is already ill-posed in `develop`, since we treat:
```python
conflicts("foo ^bar")
```
the same way as:
```python
conflicts("^foo ^bar")
```
while they're clearly not the same. Here, both to clarify the semantic and to improve speed, I took the approach that `conflicts` can be declared only on direct dependencies, and only using a `^` before the dependency. The impact on the built-in repository is small (see #38748), so I would expect the impact on custom repositories to be equally small[^1]

This is in line with the local nature of all the other directives, and allows to compute the `condition_set` atoms faster. See also this comment in the code:

https://github.com/spack/spack/blob/89e7aa990b09df631ed755aeaf676a11f803ac62/lib/spack/spack/solver/concretize.lp#L268-L277

and #38748 which are related. In the latter PR, the only use case where it is sensible to have conflicts on an arbitrary node, is to model a package that vendors another. In that case, I think we should just add another directive:
```python
vendors("foo")
```
and encode separate rules for that.

**TL:DR** What I would do, is to add `vendor` directive and implement an audit that could flag the cases where we write:
```python
conflicts("foo")
```
but we mean:
```python
conflicts("^foo")
```

[^1]: Are we aware of repositories where people have a lot of conflicts on arbitrary packages in a recipe?

alalazo replied:
Adding `^` automatically, and emitting a warning, is done in https://github.com/spack/spack/pull/38447/commits/1b90872db26fd1166b4c03cb24cd171bd970a2cf

tgamblin replied:
See further comments about this in latest review -- I think we should consolidate `vendors()` and `conflicts()`

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2399,6 +2530,18 @@ class SpecBuilder(object):
         )
     )
 
+    node_regex = re.compile(r"node\(\d,\"(.*)\"\)")
```
This shouldn't be necessary -- the functions come in properly nested from clingo; we just stringily the arguments in `extract_args`.  I think we should instead modify `extract_args` to *not* just `stringify` and to parse properly and to restore `AspFunction`s where it makes sense.

Right now this is basically stringifying then parsing again.

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/57d40f9e2542315d4a8fd73fc59a072f49aeb02f

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,31 +7,112 @@
 % This logic program implements Spack's concretizer
 %=============================================================================
 
+#const root_node_id = 0.
```
I think this is confusing b/c it's overloading the word `root`.  Is there something else we could call this?  `start_node_id`?  `base_node_id`? `initial_replica_id`?  I can't think of anything particularly good but something other than root seems needed.

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/5b6dfc346c6b71bcac71e85f363f17df5ddbdeaa

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2399,6 +2530,18 @@ class SpecBuilder(object):
+    node_regex = re.compile(r"node\(\d,\"(.*)\"\)")
+
+    @staticmethod
+    def root_node(*, pkg: str) -> str:
```
This should be renamed to something other than `root_node` (see other comment on the constant in `concretize.lp`)

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/89e7aa990b09df631ed755aeaf676a11f803ac62

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,31 +7,112 @@
-% looking for solutions to the ASP problem with "errors", which results
-% in better reporting for users. See #30669 for details.
-1 { literal_solved(ID) : literal(ID) }.
+special_case("node_flag_source").
+special_case("depends_on").
```
I know it's explained 10 lines later but I think these lines also need a comment here to say why they're a special case.    Also can this be called something like `multiple_node_args`?

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/56cd7c91f90ca608a5f5f03a061623299aa56455

tgamblin reviewed (changes requested):
@alalazo this is mostly looking awesome -- I'm planning to finish going through all the commits tomorrow but I thought I would put in some change requests now to get things started.  Most of these suggestions are around making the naming intelligible.  The logic looks good.

alalazo committed:
Simplify "node_has_variant" internal atom.

alalazo committed:
Rename atom: "facts" -> "pkg_fact"

alalazo committed:
Rename atom: "special_case" -> "multiple_nodes_attribute"

alalazo committed:
Rename const: "root_node_id" -> "main_node_id"

alalazo committed:
Rename method: "root_node" -> "main_node"

alalazo committed:
Remove the need for "node_regex"

alalazo committed:
Inline a few functions that are not needed anymore

alalazo committed:
Add "^" automatically for named conflicts that don't refer to 'this' package

See https://github.com/spack/spack/pull/38447#discussion_r1285291520

alalazo committed:
Add vendors directive

For the time being this directive prevents the vendored package
to be in the same DAG as the one vendoring it.

alalazo commented:
@tgamblin I think everything is addressed from the review.

So far I didn't add any documentation. Since the default is the same as on develop, users shouldn't see any visible change with this PR. Thus I'd leave documenting the feature to a following PR, where we could change the default concretization mode to:
```yaml
concretizer:
  duplicates:
    strategy: minimal
```
Let me know if you agree with that or if you'd like something to be part of this PR.

The other concretization mode:
```yaml
concretizer:
  duplicates:
    strategy: full
```
needs some more thoughts on how to express from configuration the properties we want to be enforced only on build dependencies (e.g. target or compiler). It seems to work fine when we hack rules into the solve, see https://github.com/spack/spack/pull/38447#issuecomment-1614372997 We also have to keep track of extensions to have this mode work properly with `py-setuptools` etc. (while "minimal" already works well)

Currently, we don't have a way to specify duplicate dependencies from a spec literal. In a literal a package can appear only once and it will be considered `node(0, Package)`.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -583,6 +602,14 @@ def stringify(sym):
     if isinstance(sym, (list, tuple)):
         return tuple(stringify(a) for a in sym)
 
+    try:
+        if sym.name == "node":
+            return NodeArgument(id=stringify(sym.arguments[0]), pkg=stringify(sym.arguments[1]))
```
Three things:

1. Method probably shouldn't be called `stringify` if it is going to also return a `NodeArgument`
2. The docstring, etc. need updating.
3. Should there be a case for this below under if `clingo_cffi`?  Doesn't seem like this is handled at all when using clingo CFFI.

alalazo replied:
> Should there be a case for this below under if clingo_cffi? Doesn't seem like this is handled at all when using clingo CFFI.

I don't think so. When the symbol is a function it always has the `.name` and the `.arguments` attributes.

alalazo replied:
Done in https://github.com/spack/spack/pull/38447/commits/807ff2cc5574cd2fe2d6222b6d7a05411913a780

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -518,8 +520,20 @@ def _execute_conflicts(pkg):
         if not when_spec:
             return
 
+        # TODO: (remove after v0.21)
```
after v0.22?

alalazo replied:
This is outdated by more recent review, but this was "after v0.21" - meaning this should happen on `develop` after v0.21 is released. Anyhow, I think this is not relevant anymore.

alalazo replied:
Outdated

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -518,8 +520,20 @@ def _execute_conflicts(pkg):
+                f"the conflict in package '{pkg.name}' on '{conflict_spec}' should "
+                f"start with a '^' sigil. Not using it is deprecated as of v0.21 and"
+                f" will be disallowed in v0.22"
+            )
```
I think we can get better attribution here -- can you not get file and line?  This stuff is *very* useful to users

alalazo replied:
This request is outdated, warning doesn't exist anymore.

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,31 +7,114 @@
 %=============================================================================
 
+% ID of the nodes in the "root" link-run sub-DAG
+#const main_node_id = 0.
```
Thought about this -- if we're calling them duplicates, I think we should stick with that. We can call them "dupes" or "duplicates".  `npm` uses similar nomenclature: https://docs.npmjs.com/cli/v6/commands/npm-dedupe.

```
#const min_dupe_id = 0
```

I don't think `main` makes sense because there isn't necessarily a "main" node.  Just difference instances of the same package in the graph. And the ids don't mean anything -- they're just numbres for uniqueness.

alalazo replied:
Done in e716dcef0c3789ae9acd40a388fc2033b860fe62

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -2537,19 +2707,23 @@ def reorder_flags(self):
                 # order is determined by the  DAG. A spec's flags come after any of its ancestors
                 # on the compile line
-                source_key = (spec.name, flag_type)
+                node = SpecBuilder.main_node(pkg=spec.name)
```
Why don't we just call this `make_node`?  I don't think `asp.py` has to know that it's id 0, right?  It doesn't care.  It'll always make nodes this way, never with another id, right?

alalazo replied:
Done in f49d656f943717e3e5c72da71eabc921cc461082 

> It'll always make nodes this way, never with another id, right?

I think so right now, but we might have to revisit when we introduce recursive spec syntax:
```
foo ^cmake@=3.14.0 ^(bar ^cmake@=3.14.1)
```

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,31 +7,114 @@
+#const direct_build = 2.
+
+% Allow clingo to create nodes
+{ attr("node", node(0..X-1, Package))         } :- max_nodes(Package, X), not virtual(Package).
+{ attr("virtual_node", node(0..X-1, Package)) } :- max_nodes(Package, X), virtual(Package).
```
same here -- stick to "dupes" as the term for multiple instances of the same package:

```suggestion
{ attr("node", node(0..X-1, Package))         } :- max_dupes(Package, X), not virtual(Package).
{ attr("virtual_node", node(0..X-1, Package)) } :- max_dupes(Package, X), virtual(Package).
```

alalazo replied:
Done in e716dcef0c3789ae9acd40a388fc2033b860fe62

tgamblin reviewed (changes requested):
A few more change requests -- I think this is nearly ready to go.

tgamblin commented on lib/spack/spack/directives.py:
```diff
@@ -903,6 +917,29 @@ def _execute_requires(pkg):
 
+@directive("vendors")
+def vendors(spec, when=None):
+    """Declares that a package has an internal copy of another package.
+
+    Currently, the effect is to forbid having the two packages in the same
+    "unification set".
+
+    Args:
+        spec: spec being vendored
+        when: optional constraint that triggers vendoring
+    """
```
I thought about this a bit and there are some complications.

There are some packages that "vendor" things that *won't* conflict with outside instances of the same library.  Some things vendor boost or header-only libraries, and they don't actually conflict with anything because they keep all the vendored code internal (not even exposed in symbols).  Other times, it does conflict, and you'd want to use a directive like this.

I think `conflicts()` is actually the better thing to use here, because it:
1. Exists already and doesn't add yet another directive
2. Doesn't imply that the package is vendored when it doesn't necessarily have to be to trigger this kind of issue.

Why don't we:
1. Make `conflicts("foo")` *and* `conflicts("^foo")` do this (i.e., forbid same unification set), or
2. Keep `conflicts("^foo")` with the behavior you've already implemented, but leave `conflicts("foo")` implemented as `vendors()` is here, and omit the warning you added? 

I think I like 2 because it gives you some flexibility between a global conflict (honestly I don't think there should be any conflicts outside a unification set) and a sub-DAG conflict.  I can see a use for both of those.

(2) would simplify things b/c it wouldn't require a deprecation.  Forbidding same unification set is pretty much equivalent to how we've always handled `conflicts("foo")` that I don't think it needs a warning, and if the package author *really* needs a sub-DAG conflict, they can add that with `conflicts("^foo")`.  Then we don't need `vendors()` or a warning, and we can avoid having to break or deprecate things in local repos.

alalazo replied:
The only non-straightforward case for this would be one where we have properties on the conflicting package:
```python
conflicts("foo@3")
```
cause in that case we have to move properties checking to that package. Let me check how easy that is. (If we have `vendors` semantic that would prevent `foo` to be in the same unification set and disregard `@3`).

alalazo replied:
Done in 7453f4c915083352ea7ce73f7da03877a7c8a299

tgamblin commented on lib/spack/spack/schema/concretizer.py:
```diff
@@ -28,6 +28,12 @@
             "unify": {
                 "oneOf": [{"type": "boolean"}, {"type": "string", "enum": ["when_possible"]}]
             },
+            "duplicates": {
+                "type": "object",
+                "properties": {
+                    "strategy": {"type": "string", "enum": ["none", "minimal", "full"]}
+                },
+            },
```
I think the strategy argument is probably going to end up being too imprecise, but this is fine for now.  If we merge this let's not commit to any syntax for the prefs *until* `0.21`.  I consider this syntax preliminary at least until we document it.

alalazo replied:
> I think the strategy argument is probably going to end up being too imprecise, [ ... ]

I agree with that. For what is worth that's why I put `strategy` unders `duplicates`. That leaves space in the future to add additional attributes to tune the strategies. I think we'll need more in particular when we'll refine `strategy:full`.

alalazo replied:
Done in 7453f4c915083352ea7ce73f7da03877a7c8a299

tgamblin commented on lib/spack/spack/solver/heuristic.lp:
```diff
@@ -0,0 +1,41 @@
+% Providers
+#heuristic attr("node", node(0, Package)) : default_provider_preference(Virtual, Package, 0), possible_in_link_run(Package). [30, true]
+%#heuristic provider(node(0, Package), node(0, Virtual)) : default_provider_preference(Virtual, Package, 0), possible_in_link_run(Package). [30, true]
+%#heuristic provider_weight(node(0, Package), node(0, Virtual), 0) : default_provider_preference(Virtual, Package, 0), possible_in_link_run(Package). [30, true]
```
We should remove commented code before this goes in.  Or , this should have at least some type of comment describing why it's commented.  Prefer the former.

alalazo replied:
Done in 27655ed3ffd720788e3acb03ec0d10f7ab193341

tgamblin commented on lib/spack/spack/solver/when_possible.lp:
```diff
@@ -0,0 +1,20 @@
+% Spack Project Developers. See the top-level COPYRIGHT file for details.
+%
+% SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
```
This and other `.lp` files need a comment at top describing what they're for

alalazo replied:
Done in 785b1d4fbd4584bb16bcc62dfb5015daefddad04

tgamblin commented on share/spack/gitlab/cloud_pipelines/stacks/e4s-oneapi/spack.yaml:
```diff
@@ -148,7 +148,7 @@ spack:
   - netlib-scalapack
   - omega-h
-  - openmpi
+  # - openmpi
```
Does reformulating `conflicts()` as I suggested (combining with `vendors()`) remove the need for this?

alalazo replied:
No. This is related to the fact that:
```yaml
spack:
  specs:
  - openmpi
  packages:
    mpi:
      require: mpich
```
is unsat in this PR, while it can be concretized on `develop`. It seems like a bug on `develop` to me. Would you agree?

tgamblin replied:
I don't think so -- the way I read that is "if the user or a package requested `mpi`, make it `mpich`".  But the user didn't request `mpi` here.  They requested `openmpi`, so there should be no issue with the provider requirement (`openmpi` doesn't require `mpi`, even if it is one).  I think it's important for the virtual constraints to *only* trigger when packages are being used *to satisfy a virtual*.

alalazo replied:
I can double check this (untested) behavior on `develop` and try replicate it 1:1. Note though that:
```yaml
spack:
  specs:
  - openmpi
  - mpich
  - hdf5
  packages:
    mpi:
      require: mpich
  concretizer:
    unify: true
```
is unsat in this PR, while on `develop` it flips the default value of the `mpi` variant for `hdf5` - so no `mpi` is used.[^1] I think the behavior in this PR is maybe less unexpected?

[^1]: `mpich` and `openmpi` can be concretized, but then no other spec can use `mpi` without triggering a conflict between the two providers.

haampie replied:
Just my two cents: my mental model for `unify: true` is an anonymous package with the `specs` list as direct deps. I wouldn't expect `specs: [mpich, openmpi]` to concretize at all with `unify:true`, and instead follow the 1 provider per DAG rule. I find the suggestions above rather unintuitive.

In practice I'd expect `unify: true` to give usable default views. With the suggestion above, you get clashes on `lib/libmpi.so` and `bin/mpicc` etc.

I'd rather see a concretization error.

If that's too strict for your taste there's always `when_possible`.

alalazo replied:
@tgamblin See https://github.com/spack/spack/pull/38447#issuecomment-1679064680 for the fix to this comment. I still think the required behavior might be unexpected, but we can fine tune that when we'll review #35322 (which I plan to do once this is merged).

tgamblin replied:
I missed @haampie's comment above but I'm not opposed to changing this semantic.  I do think it should be in a follow-on PR though, and easier to surface than a discussion inside this concretizer rework.

My worry is that there are actually a lot of things that *happen* to provide various interfaces, and there shouldn't necessarily be a conflict between them if they're requested independently.  My definition for `unify: true` is one node per package and one provider *within* a process, and there's no reason with `RPATH` semantics for roots that aren't providing `mpi` to interfere with `mpi` being used elsewhere.  But let's hash that out separately -- I think it's more important here to not break existing stacks, and we can be more explicit about potentially breaking concretizer changes with a smaller PR.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -801,6 +831,14 @@ def on_model(model):
         solve_result = self.control.solve(**solve_kwargs)
+
+        if solve_result.satisfiable and self._model_has_cycles(models):
+            warnings.warn(f"cycles detected, falling back to slower algorithm [specs={specs}]")
```
This isn't a warning -- it should just be a debug message b/c the user shouldn't have to do anything or worry about it. They'll see slower concretization time.

We *could* make it a regular message but I think "falling back to slower algorithm" is too much detail for that. Pretty convinced this belongs as a debug message.

alalazo replied:
Done in 4512233c8accdc04ca04e9c77ff63788471acd40

tgamblin commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -7,31 +7,114 @@
+     unification_set(_, ParentNode).
+
+% Any dependency of type "build" in a unification set that is not "root", stays in that unification set
+unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type == "build", SetID == "generic_build", unification_set(SetID, ParentNode).
```
can these long lines be split up? I think this file should have a column limit of 100 just like the others... we just don't have a formatter for it.

alalazo replied:
Done in 5408d69abd1fae5dabb8efaf2e7fae219e75f1f2

tgamblin reviewed (changes requested):
Last set of change requests -- I think I'm done reviewing this now (modulo iterating from here).

alalazo committed:
Rename "stringify", improve docs

alalazo committed:
Rename ""*_node" -> "*_dupe"

alalazo committed:
Rename "main_node" -> "make_node"

alalazo committed:
Demote warning to debug message

alalazo committed:
Reduce line length in lp file

alalazo committed:
Remove commented out code in lp files

alalazo committed:
Add a description at the top of lp files

alalazo committed:
Rework conflicts so that "vendors" is not needed anymore

alalazo committed:
Change "effect_rules" for symmetry with trigger rules

This even though right now we don't have cases where
the effect is on another package.

alalazo committed:
Change semantic for providers

If a possible provider is not used to satisfy a vdep,
then it's not a provider of that vdep.

alalazo commented:
The latest commit https://github.com/spack/spack/pull/38447/commits/c84df002e5f89e8e7edc372a7ebe18c68c56dc45 solves https://github.com/spack/spack/pull/38447#discussion_r1293007160 according to what we decided in DM.

Here's the build of `openmpi` in the e4s-oneapi stack: https://gitlab.spack.io/spack/spack/-/jobs/7985353

Comparing performance:

[radiuss-8a2efcf.csv](https://github.com/spack/spack/files/12345227/radiuss-8a2efcf.csv)
[radiuss-c84df00.csv](https://github.com/spack/spack/files/12345228/radiuss-c84df00.csv)
[radiuss.txt](https://github.com/spack/spack/files/12345229/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/2282ba1f-3891-4c86-a038-d7b9adcd5af7)

The change in the concretizer didn't impact performance in the benchmark.

adamjstewart commented:
Also here:
```
==> Warning: using "NodeArgument(id='0', pkg='python')@3.7.17" which is a deprecated version
```
